File: app/src/main/java/yagoc/pieces/Pieces.java
package yagoc.pieces;

import yagoc.board.BoardReader;
import yagoc.board.Move;
import yagoc.board.Square;

import java.util.Collection;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Pieces {
    public static Piece blackRook = new Rook(PieceColor.blackSet);
    public static Piece blackKnight = new Knight(PieceColor.blackSet);
    public static Piece blackBishop = new Bishop(PieceColor.blackSet);
    public static Piece blackQueen = new Queen(PieceColor.blackSet);
    public static Piece blackKing = new King(PieceColor.blackSet);
    public static Piece blackPawn = new Pawn(PieceColor.blackSet);

    public static Piece whiteRook = new Rook(PieceColor.whiteSet);
    public static Piece whiteKnight = new Knight(PieceColor.whiteSet);
    public static Piece whiteBishop = new Bishop(PieceColor.whiteSet);
    public static Piece whiteQueen = new Queen(PieceColor.whiteSet);
    public static Piece whiteKing = new King(PieceColor.whiteSet);
    public static Piece whitePawn = new Pawn(PieceColor.whiteSet);

    public static Piece none = new Piece(null, null) {
        @Override
        public boolean isValidForPiece(BoardReader board, Move move) {
            return false;
        }

        @Override
        public Stream<Move> generateMovesForPiece(BoardReader board, Square from) {
            return Stream.empty();
        }
    };

    public static Collection<Piece> all = Stream.of(
            blackRook,
            blackKnight,
            blackBishop,
            blackQueen,
            blackKing,
            blackPawn,
            whiteRook,
            whiteKnight,
            whiteBishop,
            whiteQueen,
            whiteKing,
            whitePawn,
            none)
            .collect(Collectors.toSet());

    public static Piece parse(char c) {
        switch (c) {
            case 'p':
                return whitePawn;
            case 'r':
                return whiteRook;
            case 'n':
                return whiteKnight;
            case 'b':
                return whiteBishop;
            case 'q':
                return whiteQueen;
            case 'k':
                return whiteKing;
            case 'P':
                return blackPawn;
            case 'R':
                return blackRook;
            case 'N':
                return blackKnight;
            case 'B':
                return blackBishop;
            case 'Q':
                return blackQueen;
            case 'K':
                return blackKing;
            default:
                return none;
        }
    }
}

File: app/src/main/kotlin/yagoc/pieces/Pieces.kt
package yagoc.pieces

import yagoc.board.BoardView
import yagoc.board.Move
import yagoc.board.Square
import java.util.stream.Stream

@JvmField
val blackRook: Piece = Rook(PieceColor.BlackSet)

@JvmField
val blackKnight: Piece = Knight(PieceColor.BlackSet)

@JvmField
val blackBishop: Piece = Bishop(PieceColor.BlackSet)

@JvmField
val blackQueen: Piece = Queen(PieceColor.BlackSet)

@JvmField
val blackKing: Piece = King(PieceColor.BlackSet)

@JvmField
val blackPawn: Piece = Pawn(PieceColor.BlackSet)

@JvmField
val whiteRook: Piece = Rook(PieceColor.WhiteSet)

@JvmField
val whiteKnight: Piece = Knight(PieceColor.WhiteSet)

@JvmField
val whiteBishop: Piece = Bishop(PieceColor.WhiteSet)

@JvmField
val whiteQueen: Piece = Queen(PieceColor.WhiteSet)

@JvmField
val whiteKing: Piece = King(PieceColor.WhiteSet)

@JvmField
val whitePawn: Piece = Pawn(PieceColor.WhiteSet)

@JvmField
val none: Piece = object : Piece(PieceType.None, PieceColor.None) {
    public override fun isValidForPiece(board: BoardView, move: Move): Boolean {
        throw RuntimeException("Should not happen")
    }

    public override fun generateMovesForPiece(board: BoardView, from: Square): Stream<Move> {
        throw RuntimeException("Should not happen")
    }
}

object Pieces {
    fun parse(c: Char): Piece {
        return when (c) {
            'p' -> whitePawn
            'r' -> whiteRook
            'n' -> whiteKnight
            'b' -> whiteBishop
            'q' -> whiteQueen
            'k' -> whiteKing
            'P' -> blackPawn
            'R' -> blackRook
            'N' -> blackKnight
            'B' -> blackBishop
            'Q' -> blackQueen
            'K' -> blackKing
            else -> none
        }
    }
}